### PR TITLE
Add unit testing for multikey tuplesort.

### PR DESCRIPTION
**Pull request by dkovalev1**
Add unit testing for multikey tuplesort.

There are some issues related to the multikey tuplesort which is specific 
to this version. It is tested by the regression suite, but it's difficult 
to evaluate actual test coverage and which corner cases are covered.

This commit is to extend tuplesort_mk with a specific unit tests. Tests focus on
disk-based logical tapes sorting include sort for tuples with fixed-length 
binary attributes and sorting for varlena attributes.

To have a wider coverage, the test tries different number of attributes and, 
for the varlena, used pseudo-random field length. Also the number of keys can 
be vary.

### CODE REVIEW

**Comment by RekGRpth** [comment]
```sh
pushd /home/gpadmin/src/gpdb6/src/backend/utils/sort
make CFLAGS=-DUNITTEST -s unittest-check
[          OK ] test__gp_strxfrm__LongQuery
[=============] 1 tests ran
[ PASSED      ] 1 tests
LOG:  (58P01) could not open file "postmaster.pid": No such file or directory
LOG:  Running test: readtup_heap_fail_len
LOG:  Running test: writetup_heap_fail_len
FATAL:  (XX000) could not create temporary file "base/pgsql_tmp/pgsql_tmp_test_sort_mk_sortstate": No such file or directory
          LOCATION:  OpenTemporaryFileInTablespace, fd.c:1550
          Traceback 0:  ./tuplesort_mk.t(+0x7f7283) [0x56ee6ccad283]
          Traceback 1:  ./tuplesort_mk.t(+0x7fa6a3) [0x56ee6ccb06a3]
          Traceback 2:  ./tuplesort_mk.t(+0x667890) [0x56ee6cb1d890]
          Traceback 3:  ./tuplesort_mk.t(+0x667258) [0x56ee6cb1d258]
          Traceback 4:  ./tuplesort_mk.t(+0x66b3f2) [0x56ee6cb213f2]
          Traceback 5:  ./tuplesort_mk.t(+0xcf002f) [0x56ee6d1a602f]
          Traceback 6:  ./tuplesort_mk.t(+0xcf9e92) [0x56ee6d1afe92]
          Traceback 7:  ./tuplesort_mk.t(+0xcfa242) [0x56ee6d1b0242]
          Traceback 8:  ./tuplesort_mk.t(+0xceded1) [0x56ee6d1a3ed1]
          Traceback 9:  ./tuplesort_mk.t(+0xcee242) [0x56ee6d1a4242]
          Traceback 10:  ./tuplesort_mk.t(+0xcfa318) [0x56ee6d1b0318]
          Traceback 11:  /lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x7032c7029d90]
          Traceback 12:  /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x80) [0x7032c7029e40]
          Traceback 13:  ./tuplesort_mk.t(+0x116da5) [0x56ee6c5ccda5]
[          OK ] test_tuplesort_mk_readtup_heap_fail_len
[          OK ] test_tuplesort_mk_writetup_heap_fail_len
PANIC:  (XX000) proc_exit() called in child process
          LOCATION:  proc_exit, ipc.c:104
          Traceback 0:  ./tuplesort_mk.t(+0x7f7283) [0x56ee6ccad283]
          Traceback 1:  ./tuplesort_mk.t(+0x7fa6a3) [0x56ee6ccb06a3]
          Traceback 2:  ./tuplesort_mk.t(+0x674d81) [0x56ee6cb2ad81]
          Traceback 3:  ./tuplesort_mk.t(+0x7f7729) [0x56ee6ccad729]
          Traceback 4:  ./tuplesort_mk.t(+0x7fa832) [0x56ee6ccb0832]
          Traceback 5:  ./tuplesort_mk.t(+0x667890) [0x56ee6cb1d890]
          Traceback 6:  ./tuplesort_mk.t(+0x667258) [0x56ee6cb1d258]
          Traceback 7:  ./tuplesort_mk.t(+0x66b3f2) [0x56ee6cb213f2]
          Traceback 8:  ./tuplesort_mk.t(+0xcf002f) [0x56ee6d1a602f]
          Traceback 9:  ./tuplesort_mk.t(+0xcf9e92) [0x56ee6d1afe92]
          Traceback 10:  ./tuplesort_mk.t(+0xcfa242) [0x56ee6d1b0242]
          Traceback 11:  ./tuplesort_mk.t(+0xceded1) [0x56ee6d1a3ed1]
          Traceback 12:  ./tuplesort_mk.t(+0xcee242) [0x56ee6d1a4242]
          Traceback 13:  ./tuplesort_mk.t(+0xcfa318) [0x56ee6d1b0318]
          Traceback 14:  /lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x7032c7029d90]
          Traceback 15:  /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x80) [0x7032c7029e40]
          Traceback 16:  ./tuplesort_mk.t(+0x116da5) [0x56ee6c5ccda5]
make[1]: *** [../../../../../src/Makefile.mock:25: tuplesort_mk-check] Aborted (core dumped)
make: *** [../../../../src/backend/common.mk:52: unittest-check-local] Error 2
```

File: src/backend/utils/sort/test/tuplesort_mk_test.c [hunk 1]
```diff
@@ -8,6 +8,41 @@
...
+#include "storage/proc.h"
+
+#include <sys/stat.h>
+
+#define get_compare_function_for_ordering_op mock_get_compare_function_for_ordering_op
+#define ScanKeyEntryInitialize mock_ScanKeyEntryInitialize
+
+/*
+	Modify the following to check more possible cases
+*/
```
Comment (RekGRpth) [inline src/backend/utils/sort/test/tuplesort_mk_test.c:21]:
```suggestion
 * Modify the following to check more possible cases
 */
```

File: src/backend/utils/sort/test/tuplesort_mk_test.c [hunk 2]
```diff
@@ -203,17 +286,380 @@ test_tuplesort_mk_writetup_heap_fail_len(void **test_state)
...
+test_sort_varlena(void **stateptr)
+{
+	int nkeys, nattrs;
+
+
+	for (nattrs = MIN_ATTRS; nattrs <= MAX_ATTRS; nattrs++)
+		for (nkeys = MIN_KEYS; nkeys <= Min(nattrs, MAX_KEYS); nkeys++)
+		{
+			run_sort_test_varlena(nattrs, nkeys);
+		}
```
Comment (RekGRpth) [inline src/backend/utils/sort/test/tuplesort_mk_test.c:623]:
```suggestion
	for (int nattrs = MIN_ATTRS; nattrs <= MAX_ATTRS; nattrs++)
		for (int nkeys = MIN_KEYS; nkeys <= Min(nattrs, MAX_KEYS); nkeys++)
			run_sort_test_varlena(nattrs, nkeys);
```

File: src/backend/utils/sort/test/tuplesort_mk_test.c [hunk 3]
```diff
@@ -203,17 +286,380 @@ test_tuplesort_mk_writetup_heap_fail_len(void **test_state)
...
+	int nattrs = 2;
+
+	mkdir("base", S_IRWXU);
+
+	for (nattrs = MIN_ATTRS; nattrs <= MAX_ATTRS; nattrs++)
+		for (nkeys = MIN_KEYS; nkeys <= Min(nattrs, MAX_KEYS); nkeys++)
+		{
+			// printf("nkeys: %d, nattrs: %d\n", nkeys, nattrs);
+			run_sort_test_fixed(nattrs, nkeys);
+		}
```
Comment (RekGRpth) [inline src/backend/utils/sort/test/tuplesort_mk_test.c:609]:
```suggestion
	mkdir("base", S_IRWXU);

	for (int nattrs = MIN_ATTRS; nattrs <= MAX_ATTRS; nattrs++)
		for (int nkeys = MIN_KEYS; nkeys <= Min(nattrs, MAX_KEYS); nkeys++)
			run_sort_test_fixed(nattrs, nkeys);
```

File: src/backend/utils/sort/test/tuplesort_mk_test.c [hunk 4]
```diff
@@ -203,17 +288,376 @@ test_tuplesort_mk_writetup_heap_fail_len(void **test_state)
...
+		if (TupIsNull(slot))
+			break;
+
+		HeapTuple tuple = ExecFetchSlotHeapTuple(slot);
+		heap_deform_tuple(tuple, tupdesc, values, isnull);
+
+		// int v0 = DatumGetInt32(values[0]);
+		// int v1 = DatumGetInt32(values[1]);
+		
+		// printf("(%d,%d)", v0, v1);
```
Comment (RekGRpth) [inline src/backend/utils/sort/test/tuplesort_mk_test.c:414]: dead code?

File: src/backend/utils/sort/test/tuplesort_mk_test.c [hunk 5]
```diff
@@ -203,17 +288,376 @@ test_tuplesort_mk_writetup_heap_fail_len(void **test_state)
...
+		pfree(values);
+		pfree(isnull);
+    }
+	// printf("\n");
```
Comment (RekGRpth) [inline src/backend/utils/sort/test/tuplesort_mk_test.c:418]: dead code?

File: src/backend/utils/sort/test/tuplesort_mk_test.c [hunk 6]
```diff
@@ -203,17 +288,368 @@ test_tuplesort_mk_writetup_heap_fail_len(void **test_state)
...
+    MemoryContext mcxt = AllocSetContextCreate(CurrentMemoryContext,
+                                               "mk_test",
+                                               ALLOCSET_DEFAULT_SIZES);
+
+
+	MemoryContext old_cxt = MemoryContextSwitchTo(mcxt);
+
+	FormData_pg_attribute *attributes = (FormData_pg_attribute *)palloc0(sizeof(FormData_pg_attribute) * nattrs);
+	Form_pg_attribute *attrs = (Form_pg_attribute *)palloc0(sizeof(Form_pg_attribute) * nattrs);
+	for (i = 0; i < nattrs; i++)
```
Comment (RekGRpth) [inline src/backend/utils/sort/test/tuplesort_mk_test.c:305]:
```suggestion
	for (int i = 0; i < nattrs; i++)
```

File: src/backend/utils/sort/test/tuplesort_mk_test.c [hunk 7]
```diff
@@ -203,17 +288,368 @@ test_tuplesort_mk_writetup_heap_fail_len(void **test_state)
 	assert_true(error_thrown);
 }
 
+void run_sort_test_fixed(int nattrs, int nkeys)
+{
+	int i;
```
Comment (RekGRpth) [inline src/backend/utils/sort/test/tuplesort_mk_test.c:293]:
```suggestion
```

File: src/backend/utils/sort/test/tuplesort_mk_test.c [hunk 8]
```diff
@@ -203,17 +288,368 @@ test_tuplesort_mk_writetup_heap_fail_len(void **test_state)
...
+		len = strlen(buffer);
+	}
+
+	text *ret_text = cstring_to_text(buffer);
+	return ret_text;
+}
+
+void run_sort_test_varlena(int nattrs, int nkeys)
+{
+	int i;
```
Comment (RekGRpth) [inline src/backend/utils/sort/test/tuplesort_mk_test.c:455]:
```suggestion
```

File: src/backend/utils/sort/test/tuplesort_mk_test.c [hunk 9]
```diff
@@ -8,6 +8,43 @@
...
+#define MAX_TEST_STRING_LEN 1024
+
+void run_sort_test_fixed(int nattrs, int nkeys);
+void run_sort_test_varlena(int nattrs, int nkeys);
+text *make_test_string(int idx);
+int random_sequence(int key);
+Datum test_textcmp(PG_FUNCTION_ARGS);
+
+void
+mock_ScanKeyEntryInitialize(ScanKey entry,
```
[the remark this answers is not in the document]
Comment (dkovalev1) [reply src/backend/utils/sort/test/tuplesort_mk_test.c]: this is a main module, not a library, and static here is not so important, but, okay, why not

**Review by RekGRpth — APPROVED** [review]

File: src/backend/utils/sort/test/tuplesort_mk_test.c [hunk 10]
```diff
@@ -27,6 +64,54 @@ struct LogicalTapeSet
...
+	return true;
+}
+
+static Datum
+test_textcmp(PG_FUNCTION_ARGS)
+{
+	text*		a = DatumGetTextP(PG_GETARG_DATUM(0));
+	text*		b = DatumGetTextP(PG_GETARG_DATUM(1));
+
+	PG_RETURN_INT32(strcmp(text_to_cstring(a), text_to_cstring(b)));
```
Comment (KnightMurloc) [inline src/backend/utils/sort/test/tuplesort_mk_test.c:81]: Is it fine that memory is leaking here?

File: src/backend/utils/sort/test/tuplesort_mk_test.c [hunk 11]
```diff
@@ -27,6 +64,54 @@ struct LogicalTapeSet
...
+	*cmpfunc = InvalidOid;
+	*reverse = false;
+	return true;
+}
+
+static Datum
+test_textcmp(PG_FUNCTION_ARGS)
+{
+	text*		a = DatumGetTextP(PG_GETARG_DATUM(0));
+	text*		b = DatumGetTextP(PG_GETARG_DATUM(1));
```
Comment (KnightMurloc) [inline src/backend/utils/sort/test/tuplesort_mk_test.c:79]:
```suggestion
	text*		a = PG_GETARG_TEXT_PP(0);
	text*		b = PG_GETARG_TEXT_PP(1);
```

File: src/backend/utils/sort/test/tuplesort_mk_test.c [hunk 12]
```diff
@@ -75,10 +75,18 @@ mock_get_compare_function_for_ordering_op(Oid opno, Oid *cmpfunc, bool *reverse)
...
+	text*		b = PG_GETARG_TEXT_PP(1);
 
-	PG_RETURN_INT32(strcmp(text_to_cstring(a), text_to_cstring(b)));
+	char *pa = text_to_cstring(a);
+	char *pb = text_to_cstring(b);
+
+	int ret = strcmp(pa, pb);
+
+	pfree(pa);
+	pfree(pb);
```
[the remark this answers is not in the document]
Comment (dkovalev1) [reply src/backend/utils/sort/test/tuplesort_mk_test.c]: Good point!

**Review by RekGRpth — APPROVED** [review]

**Review by KnightMurloc — APPROVED** [review]
